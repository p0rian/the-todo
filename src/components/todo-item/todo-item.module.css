.itemStyle {
  display: flex;
  align-items: center;
  margin: 10px;
  padding: 10px;
  border-radius: 4px;
  background-color: #fff;
  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.15);
}

.itemStyle:hover {
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.15);
}

.checkboxStyle {
  margin-right: 10px;
}

.inputStyle {
  margin-right: 10px;
}

.editButtonsStyle {
  display: flex;
  align-items: center;
  margin-left: auto;
}

.removeButtonStyle {
  color: red;
}
