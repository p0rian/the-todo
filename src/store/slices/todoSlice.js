import { createSlice, createEntityAdapter } from "@reduxjs/toolkit";
import { v4 as uuidv4 } from "uuid";

const todosAdapter = createEntityAdapter({});
const sliceName = "todo";
const todoSlice = createSlice({
  name: sliceName,
  initialState: todosAdapter.getInitialState(),
  reducers: {
    addTodo: {
      reducer: (state, action) => {
        const { id, createdOn, title, isCompleted } = action.payload;
        todosAdapter.addOne(state, { id, createdOn, title, isCompleted });
      },
      prepare: (title) => {
        return {
          payload: {
            id: uuidv4(),
            createdOn: Date.now(),
            title,
            isCompleted: false,
          },
        };
      },
    },
    updateTodo: todosAdapter.updateOne,
    removeTodo: todosAdapter.removeOne,
  },
});

export const { addTodo, updateTodo, removeTodo } = todoSlice.actions;

export const { selectAll: selectAllTodos, selectById: selectTodoById } =
  todosAdapter.getSelectors((state) => state.todoData);

export const todoReducer = todoSlice.reducer;
