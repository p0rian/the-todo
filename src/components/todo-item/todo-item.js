import { Checkbox } from "@mui/material";
import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { removeTodo, updateTodo } from "store/slices/todoSlice";
import { Button, Input, Box } from "@mui/material";

import css from "./todo-item.module.css";

export const TodoItem = ({ item }) => {
  const [editing, setEditing] = useState(false);
  const [editedTitle, setEditedTitle] = useState(item.title);
  const dispatch = useDispatch();

  const handleRemove = () => {
    dispatch(removeTodo(item.id));
  };

  const handleStartEditing = () => {
    setEditing(true);
  };

  const handleFinishEditing = () => {
    if (editedTitle.trim() !== "") {
      dispatch(
        updateTodo({
          id: item.id,
          changes: {
            title: editedTitle,
          },
        })
      );
    }
    setEditing(false);
  };

  const handleCancelEditing = () => {
    setEditedTitle(item.title);
    setEditing(false);
  };

  const handleInputChange = (e) => {
    setEditedTitle(e.target.value);
  };

  const handleCompletedStatus = (e) => {
    dispatch(
      updateTodo({
        id: item.id,
        changes: {
          isCompleted: !item.isCompleted,
        },
      })
    );
  };

  return (
    <li>
      {editing ? (
        <div className={css.itemStyle}>
          <Input type="text" value={editedTitle} onChange={handleInputChange} />
          <Button onClick={handleFinishEditing}>Save</Button>
          <Button onClick={handleCancelEditing}>Cancel</Button>
        </div>
      ) : (
        <div className={css.itemStyle}>
          <Checkbox
            checked={item.isCompleted}
            onChange={handleCompletedStatus}
          ></Checkbox>
          <Box component="span" onClick={handleStartEditing}>
            {item.title}
          </Box>
          <Button onClick={handleRemove}>X</Button>
        </div>
      )}
    </li>
  );
};
