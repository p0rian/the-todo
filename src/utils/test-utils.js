/* eslint-disable react/prop-types */
import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { todoReducer } from "store/slices/todoSlice";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: { todoData: todoReducer }, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
