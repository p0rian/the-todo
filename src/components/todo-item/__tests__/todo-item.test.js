import React from "react";
import { renderWithProviders } from "utils/test-utils";
import { updateTodo } from "store/slices/todoSlice";
import { useDispatch } from "react-redux";
import { screen, fireEvent } from "@testing-library/react";
import { TodoItem } from "../todo-item";

describe("TodoItem", () => {
  const testItem = {
    id: "testId",
    title: "Test title",
    isCompleted: false,
  };

  it("renders correctly", () => {
    const { container } = renderWithProviders(<TodoItem item={testItem} />);
    expect(container).toMatchSnapshot();
  });
  it("renders the item title and delete button", () => {
    renderWithProviders(<TodoItem item={testItem} />);

    expect(screen.getByText(testItem.title)).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "X" })).toBeInTheDocument();
  });

  it("marks the item as completed when checkbox is checked", () => {
    renderWithProviders(<TodoItem item={testItem} />);

    fireEvent.click(screen.getByRole("checkbox"));
    expect(screen.getByRole("checkbox")).toBeChecked();
  });

  // it("opens edit mode when clicking on the item title", () => {
  //   renderWithProviders(<TodoItem item={testItem} />);

  //   fireEvent.click(screen.getByText(testItem.title));
  //   expect(screen.getByRole("textbox")).toBeInTheDocument();
  //   expect(screen.getByRole("button", { name: "Save" })).toBeInTheDocument();
  //   expect(screen.getByRole("button", { name: "Cancel" })).toBeInTheDocument();
  // });

  // it("updates the item title when 'Save' is clicked", () => {
  //   const { store } = renderWithProviders(<TodoItem item={testItem} />);

  //   fireEvent.click(screen.getByText(testItem.title));
  //   fireEvent.change(screen.getByRole("textbox"), { target: { value: "New title" } });
  //   fireEvent.click(screen.getByRole("button", { name: "Save" }));
  //   expect(store.getState().todoData.entities[testItem.id].title).toBe("New title");
  // });

  // it("cancels edit mode when 'Cancel' is clicked", () => {
  //   renderWithProviders(<TodoItem item={testItem} />);

  //   fireEvent.click(screen.getByText(testItem.title));
  //   fireEvent.change(screen.getByRole("textbox"), { target: { value: "New title" } });
  //   fireEvent.click(screen.getByRole("button", { name: "Cancel" }));
  //   expect(screen.queryByRole("textbox")).not.toBeInTheDocument();
  //   expect(screen.getByText(testItem.title)).toBeInTheDocument();
  // });

  // it("deletes the item when 'X' is clicked", () => {
  //   const { store } = renderWithProviders(<TodoItem item={testItem} />);

  //   fireEvent.click(screen.getByRole("button", { name: "X" }));
  //   expect(store.getState().todoData.entities[testItem.id]).toBeUndefined();
  // });
});
