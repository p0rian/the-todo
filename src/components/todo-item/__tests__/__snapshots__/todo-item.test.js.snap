// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem renders correctly 1`] = `
<div>
  <li>
    <div
      class="itemStyle"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-root MuiCheckbox-colorPrimary css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiBox-root css-0"
      >
        Test title
      </span>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        X
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </li>
</div>
`;
